ssl/statem/statem_dtls.c

```diff
@@ -818,10 +818,36 @@ static int dtls1_process_out_of_seq_message(SSL_CONNECTION *s,
     return 0;
 }
 
+static int dtls1_read_hm_header(unsigned char *msgheaderstart,
+                                struct hm_header_st *msg_hdr)
+{
+    unsigned long msg_len, frag_off, frag_len;
+    PACKET msgheader;
+
+    if (!PACKET_buf_init(&msgheader, msgheaderstart, DTLS1_HM_HEADER_LENGTH)
+            || !PACKET_get_1(&msgheader, (unsigned int *)&msg_hdr->type)
+            || !PACKET_get_net_3(&msgheader, &msg_len)
+            || !PACKET_get_net_2(&msgheader, (unsigned int *)&msg_hdr->seq)
```

Comment:
Both casts are wrong. You must  add a temporary unsigned int variable for both type and seq otherwise you're overwriting other struct members of msg_hdr and the value will be a completely different value on big endian.

OTOH please note there is PACKET_get_net_3_len, which has size_t data argument which can be used directly without the temporary unsigned long values  and without the SIZE_MAX checks as we do not support any architectures where sizeof(size_t) < 3.

Reply:
Good point. Thanks!